### PR TITLE
Support individual rule options

### LINKED ISSUE #69: Support individual rule options
Depends on #67

Currently, the rule's `check` function has a signature `check(root: Folder, context?: Context): Promise<Array<Diagnostic>>`. This is limiting our ability to reasonably pass options to rules, so instead we should revise this to make the second optional argument be the options, if the rule requires them.

The context should be scrapped. It's currently only used to determine if the source file extension is `js` or `ts`, and that is not even implemented yet. Context should be passed to rules through `this` by binding them upon the plugin creation. That way, only the plugin developer is responsible for what appears in the context. The functionality of exposing certain services to plugin writers will probably be implemented through functions that plugin writers can import.

### PR DESCRIPTION
**Pull request by daniilsapa**
Resolves #69 

Hi @illright 

I implemented full support for individual rule options.

Here are my comments:
- I looked for a candidate rule to implement first individual options for and decided to go with the threshold in shared-lib-grouping
- Changed the way we pass context to rules (but it is still provided by Steiger temporarily)

Questions:
Should the Rule interface have an optional `validateOptions` method, so a rule can validate its individual options before running to handle unexpected cases like this: 

```javascript
export default [
  ...fsd.configs.recommended,
  {
    rules: {
      'fsd/shared-lib-grouping': ['error', { threshold: 'asdf' }]
    }
  }
];
```

### CODE REVIEW

File: packages/types/index.ts [hunk 1]
```diff
@@ -13,10 +13,23 @@ export interface Context {
   sourceFileExtension: 'js' | 'ts'
 }
 
+export interface Options {
+  /** Overrides for layer names */
+  layerNames: Record<string, string>
+  additionalSegmentNames: Array<string>
+}
```
Comment (illright) [inline packages/types/index.ts:20]: issue: this is FSD-specific stuff, I don't think it belongs to Steiger core

File: packages/steiger-plugin-fsd/src/shared-lib-grouping/index.ts [hunk 2]
```diff
@@ -1,13 +1,18 @@
...
-import type { Diagnostic, Rule } from '@steiger/types'
+import type { Diagnostic, Rule, RuleOptions } from '@steiger/types'
 import { NAMESPACE } from '../constants.js'
 
-const THRESHOLD = 15
+const DEFAULT_THRESHOLD = 15
+
+export interface SharedLibGroupingOptions extends RuleOptions {
+  threshold?: number
+}
```
Comment (illright) [inline packages/steiger-plugin-fsd/src/shared-lib-grouping/index.ts:9]: issue: I understand that you want to have demo of the rule options capability, I just think that this rule shouldn't be configurable like that, not yet, anyway. Maybe later, after we do some research on the results of this rule.

suggestion: let's remove the configurability of this rule. As for the demo, I don't think we must have it no matter what, I'd prefer if we give it some separate thought.

**Review by illright — CHANGES_REQUESTED** [review]
praise: oof this is a difficult one, my head melted reviewing, I can only imagine how it must have been to write it, great job!

I feel like we might need to iterate on this PR a couple times, I tried to write as many of my thoughts as I could but there is still more. If you disagree with anything, btw, feel free to let me know, we can work out the final design together as we go

File: packages/types/index.ts (hunk 1 above)
Comment (daniilsapa) [reply packages/types/index.ts:20]: Yes, you're right. No problem, I'll remove it. I added that as a placeholder anyway, they are not supported yet.

File: packages/types/index.ts [hunk 3]
```diff
@@ -70,8 +83,7 @@ export interface ConfigObject {
   ignores?: Array<string>
   /** Severity of rules and individual rule options. */
   rules?: {
-    // TODO: add "| [Severity, Record<string, unknown>]" when individual rule options are supported
-    [ruleName: string]: Severity
+    [ruleName: string]: Severity | [Severity, Record<string, unknown>]
```
[the remark this answers is not in the document]
Comment (daniilsapa) [reply packages/types/index.ts:86]: Yes, sure, makes sense. Just missed that, thanks for catching

File: packages/types/index.ts [hunk 4]
```diff
@@ -13,10 +13,23 @@ export interface Context {
   sourceFileExtension: 'js' | 'ts'
 }
```
[the remark this answers is not in the document]
Comment (daniilsapa) [reply packages/types/index.ts:14]: Got it, ok

File: packages/steiger-plugin-fsd/src/shared-lib-grouping/index.ts (hunk 2 above)
Comment (daniilsapa) [reply packages/steiger-plugin-fsd/src/shared-lib-grouping/index.ts:9]: No problem, its main purpose was to let me manually test the new functionality and confirm that it works e2e. Then I decided to button up and commit it.

File: packages/types/index.ts [hunk 5]
```diff
@@ -13,10 +13,23 @@ export interface Context {
...
 export interface Rule {
   /** Short code name for the rule. */
   name: string
-  check: (root: Folder, context: Context) => RuleResult | Promise<RuleResult>
+  check: (
+    this: Context | Rule,
+    root: Folder,
+    options?: Options,
+    ruleOptions?: RuleOptions,
+  ) => RuleResult | Promise<RuleResult>
```
[the remark this answers is not in the document]
Comment (daniilsapa) [reply packages/types/index.ts]: Yeah, it looks like I perceived your example from the screenshot below as a call to action to add support for plugin-level options. That's why I added that Options interface and passed it as the second argument to `Rule.check`

Screenshot 2024-08-18 at 23 37 35

[the remark this answers is not in the document]
Comment (daniilsapa) [reply packages/types/index.ts]: And yes, I added generics for Rule

**Review by illright — APPROVED** [review]
Actually, never mind, it's pretty solid. The only thing that is still bugging me a little bit is the `.getState()` call. I feel like I want to create a clean data pipeline, to go all in on Effector, but we don't need to worry about it right now. Great PR!

**Pull request merged**